Implement jax.scipy.stats.truncnorm

adrn opened the pull request:
Following from #12618, here is an implementation of the logpdf, logcdf, and logsf (and un-log versions) for `scipy.stats.truncnorm`. Much of this is ported over from scipy itself, but I had to change some logic to use `select()` statements. I was a little confused by the test structure but think I figured out a solution to enforce that the upper limit parameter `b` is always larger than the lower limit `a`, and the evaluating values `x` are between these values.

adrn committed:
implement truncnorm in jax.scipy.stats

fix some shape and type issues

import into namespace

imports into non-_src library

working logpdf test

cleanup

working tests for cdf and sf after fixing select

relax need for x to be in (a, b)

ensure behavior with invalid input matches scipy

remove enforcing valid parameters in tests

added truncnorm to docs

whoops alphabetical

fix linter error

fix circular import issue

adrn commented:
@jakevdp I don't have privileges to request a review from you, so pinging you instead! (and I just filled out the CLA)

jakevdp commented on tests/scipy_stats_test.py:
```diff
@@ -472,6 +472,143 @@ def args_maker():
+      # hack to ensure that b > a and a < x < b
+      a = -np.abs(a).astype(a.dtype)
+      x = ((a + b) / 2).astype(x.dtype)
+      b = np.abs(b).astype(b.dtype)
```
What happens if these are not in order `a < x < b`? It should return `-inf`, correct? Should we just let the values be out of order so as to test that?

If we need them in order, it might be simpler to do something like this, and then cast to the desired dtype:
```python
a, x, b = rng((3, *shape), dtype).sort(0)
```

adrn replied:
Yea, true, I guess we can relax the need for `x` to be in bounds, but `b` must be greater than `a`. But since `a` or `b` can have different shapes, I couldn't figure out a clean way to use `np.sort()` to get this to work. My hack might be the simplest way of enforcing this, but I could think more about it!

jakevdp replied:
It looks like when `a >= b` the scipy version returns `nan`:
```python
from scipy.stats import truncnorm
print(truncnorm.logpdf(1, a=2, b=0))
# nan
```
I believe `_CheckAgainstNumpy` will check the locations of NaN values, so it might be best in the tests to check the full range of values to ensure that we're matching that behavior.

jakevdp reviewed (commented):
Looks great! Just a quick comment on tests.

adrn commented:
@jakevdp Sorry for the delay here! I removed the hack parameter value enforcement from the tests (now it tests invalid parameter values as well) and rebased because of a conflict.

jakevdp reviewed (commented):
Thanks for circling back to this - looks great! Two last things before we merge:

- please add the new functionality to the HTML docs in https://github.com/google/jax/blob/main/docs/jax.scipy.rst
- please squash all your changes into a single commit (see [single-change commits](https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests))

Thanks!

adrn commented:
Cool - thanks for the review! I just squashed the commits and added it to the docs.

adrn commented:
Fixed a small docs issue and pushed again, so tests might need to be approved again.

jakevdp commented:
The doc build is failing in a strange way, that seems to be related to our lazy loading of scipy. I'm not sure what the fix might be...

jakevdp commented on jax/_src/scipy/stats/truncnorm.py:
```diff
@@ -0,0 +1,129 @@
+from jax._src.numpy import lax_numpy as jnp
+from jax._src.numpy.util import _wraps
+from jax._src.numpy.lax_numpy import _promote_args_inexact
+from jax.scipy import special, stats
```
I think this is the issue: importing `stats` here leads to a circular import.

Using `from jax._src.scipy.stats import norm` should fix this.

adrn replied:
Got it - done!

jakevdp commented:
Thanks for the PR!

adrn commented:
Thanks for the review!
